### ISSUE TITLE
Mod addons

### ISSUE THREAD

**Issue by nrb547** [multitheftauto/mtasa-blue#1334]
**Is your feature request related to a problem? Please describe.**
Many SAMP players don't want to switch over from SAMP to MTA due to modding opportunities. SkyGFX, Project 2DFX, Widescreen Fix and more are frequently used modifications that allow the game to be more enjoyable. It's a pity we couldn't have any of these mods available for now, eventhough they are very popular and often used by players outside of MTA.

**Describe the solution you'd like**
I would like you to introduce addons that can be enabled in the options located in the MTA menu. Some popular mods that are available open source and don't affect the gameplay but just the looks, should have their correspondent addons. Remember that _**you**_ control the source code, not the players, so this gets rid of any cheating possibilities. The only thing a player can modify is for example the mods config, let's say you use SkyGFX so you can switch between PC, PS2 and Xbox colors.

**Describe alternatives you've considered**
There is no actual better alternative that I could think of. Letting people install whatever they want sounds too risky, maybe you could make it possible to install .asi/.dll mods if you have set it enabled as a server owner.

**Additional context**
Some have already said we don't need this, as everything can be done with Lua. First off no, you cannot do everything in Lua what you could do with native programming. Second, it has to be server-sided and an actual server owner won't care about implementing SkyGFX or anything, it would be too time consuming for him, non rewarding as in earning possibilities as they are visual mods.

**Comment by PlatinMTA** [comment]
Related forum topic: https://forum.mtasa.com/topic/123154-mods-for-mta

**Comment by qaisjp** [comment]
Thank you for the detailed issue - it's really useful having this context.

I think if we did something like this, we could also expose a server API for locking, enabling or disabling these 'addons'. 

> maybe you could make it possible to install .asi/.dll mods if you have set it enabled as a server owner

Since this was filed undo 'alternatives', I'm sure you're aware that we can't let this happen as it's susceptible to cheating + then people would look to us.

Although theoretically we could allow the server to provide a list of allowed sha256 hashes... but that's a recipe for disaster. Servers might tell unsuspecting users to download their addon, and bam, you've got some sort of virus. (They could do this anyway, but at least it wouldn't be our fault.)

----

Generally I'm hesitant about this feature, but I understand that if we want to win over more people, we might want to be a little bit more open to client customisation.

The scenario you've outlined about server owners seems quite frustrating:

> Some people told that these features can be scripted, but in reality if you're a generic player and you'll ask a server owner to add them, he will most likely be not bothered by it, as people are more busy with adding some GTA 6 next-gen GUI to this 16 year old game.

If we seriously consider this proposal, we should proceed slowly -- with caution.

**Comment by PlatinMTA** [comment]
> Servers might tell unsuspecting users to download their addon, and bam, you've got some sort of virus.

That could happend already, there isn't really that much that MTA could do to stop it apart from warning users to not download things from servers they dont trust. Even if by adding this feature (addon) it would be possible to not let users inside your server unless they have a certain addon, a server owner that asks a user that likes the server to download X, then its likely that the user would download it and get a virus if that is the intention of the server owner. 

Maybe not letting the server owners know which addons they have ON and just not letting them join if they have one addon that isn't on the list of allowed ones its enough.

I think that MTA would benefit from letting servers choose whenever they want to let users join with custom .dlls and even .asi or not. Something like a warning when you are trying to join a server which lets users use this type of modifications could be added too. The whitelist from addons it's all thats needed.

------

I don't know how the anticheat system works tho. It appears that it's more global that I thought. For example you can disable the anticheat that checks for trainers, but MTA could ban the client even if you do this. Tbh I don't see who is benefiting for this. That client wouldn't be able to join other server that has this anticheat enabled so, what's the point on banning him if he didn't try to join a server with the anticheat on?

I never tried it tho, but I just follow what the wiki says about this stuff: https://wiki.multitheftauto.com/wiki/Anti-cheat_guide

**Comment by nrb547** [comment]
Sure you can let servers decide if players can join with these addons or not, but as I pointed out they are visual mods, so servers shouldn't really care. You can literally take a stable build of an open sourced mod and build an addon for it to be available in the options.

**Comment by ghost** [comment]
> 
> I think if we did something like this, we could also expose a server API for locking, enabling or disabling these 'addons'.
> 

Yeah, I've been thinking about this as well. It can work well, but this will be more work for us. We will have to review their DLL files every time they make a change. Other than that, this would mean people can implement amazing features that might not be possible with the API we have exposed via Lua.

> Although theoretically we could allow the server to provide a list of allowed sha256 hashes... but that's a recipe for disaster. Servers might tell unsuspecting users to download their addon, and bam, you've got some sort of virus. (They could do this anyway, but at least it wouldn't be our fault.)

If we have a access to their source code, the ability to allow it or dismiss the changes they make to their module, then it might work well, but we should still warn users that we will not take any responsibility for the damage caused by the custom module. Like whenever a user tries to join a server, a warning should be displayed that the server wants the client to download and run a custom module, and it can potentially harm the client if it's malicious. The client can add exceptions in settings to disable the warning for trusted servers.

In my opinion, this can only work if we create a platform where server owners can post their custom modules for us to review along with the source code. If not, then it's bound to fail because these would leave the doors open for hackers.

**Comment by nrb547** [comment]
I don't see a reason why a server would like you to download a custom module. PS2 effects, light coronas are entirely up to the player, it just doesn't have any influence on the gameplay at all.

This is true saml1er, if you want to have the newest version possible you'd have to check regularly if there is a newer version available and make an addon upon that. But in this case it's pretty obvious that stability and security is the most important, so just take a stable, even older, build.

**Comment by ghost** [comment]
> 
> 
> I don't see a reason why a server would like you to download a custom module. PS2 effects, light coronas are entirely up to the player, it just doesn't have any influence on the gameplay at all.

About downloading a custom mod, it was just a suggestion. I was thinking of allowing server owners to use C++. That's not really important, and it's off-topic. 

Here's the problem, most of these mods are incompatible with MTA. We frequently add new features to MTA. If we try to modify anything related to rendering, these mods won't work on MTA. For the new code, I want to hook functions directly and replace them completely. If the mods are patching default SA functions, the mods will have no affect on MTA. About skygfx, I highly doubt if it's even compatible with MTA because we are directly setting render states in many parts of the code. 

SAMP doesn't receive updates as frequently as MTA. This is why most mods are compatible with SAMP. Trying not to break mods will be extra work for us. 

The only and proper way to get everything to work together is if someone writes a mod specifically for MTA. But then, what's the point? If someone wants a feature to be implemented, they can always send us a PR, and we'll happily merge it.

**Comment by qaisjp** [comment]
The use case I'm in support of is "the server owner is too lazy to add this resource / doesn't want to enable it for everyone, but doesn't mind if I have that addon".

I'm very much not in support of "C++ resources". Supporting an API for that will be a massive pain (see: server module SDK) and we don't have the _resources_ to review their code.

----

I was more thinking that _we_ integrate and distribute specific addons, and allow the user to enable or disable them in the Settings menu.

Servers (via script) could enable, disable or _lock_ the list. Locking prevents the user from enabling or disabling addons whilst ingame. But neither servers or users would be able to add new addons -- these are compiled right into MTA.

The key requirement is that the mod we integrate would have to be open source.

But like @saml1er said:

> Here's the problem, most of these mods are incompatible with MTA. 

----

But the entire idea also opens the floodgates for people wanting their favourite mod to be integrated as an addon... which will be annoying.

**Comment by PlatinMTA** [comment]
I don't think that MTA should manage the addons, that will just be a pain in the ass. Maybe an update can break the addon, and you need to fix it because it depends on the MTA Team. That would be annoying, and really I don't think that it will fix the core issue: users that come from SAMP want to use cleo mods, asi and dll modifications just like they do in SAMP. Adding addons to MTA won't fix that.

But that's just my opinion.

**Comment by nrb547** [comment]
> I was more thinking that _we_ integrate and distribute specific addons, and allow the user to enable or disable them in the Settings menu.

Exactly what I think of.

> Servers (via script) could enable, disable or _lock_ the list. Locking prevents the user from enabling or disabling addons whilst ingame. But neither servers or users would be able to add new addons -- these are compiled right into MTA.

Yes, they are compiled right into MTA. Just keep in mind to leave a small window for the config file. Mods like SkyGFX have config files, changing from PS2 to Xbox pipeline etc. should be a thing.

> Here's the problem, most of these mods are incompatible with MTA.

People run the mods at their own risk, if it works for them, fine, if not it's not really your fault. For now we had the possibility to replace files in .img archives, if a custom .dff crashes someone's game it's their fault, not yours.

> But the entire idea also opens the floodgates for people wanting their favourite mod to be integrated as an addon... which will be annoying.

As I said there are some really popular mods that are liked by hundreds of people. If you can integrate some of them, it would be just fine.

**Comment by qaisjp** [comment]
> People run the mods at their own risk, if it works for them, fine, if not it's not really your fault.

I think the key point here is that we can't integrate them (or there's no point integrating them) if they simply don't work

But I guess this can be evaluated on a case-by-case basis.

**Comment by nrb547** [comment]
Any update on this?
